data_steward/cdr_cleaner/cleaning_rules/ehr_submission_data_cutoff.py

```diff
@@ -0,0 +1,240 @@
+"""
+Enforces a data cutoff date for PPI data.
+
+Original Issue: DC-1445
+
+Intent is to enforce the data cutoff date for PPI data in all CDM tables excluding the person table by sandboxing and
+    removing any records that persist after the data cutoff date.
+"""
+
+# Python imports
+import logging
+from datetime import datetime
+
+# Project imports
+from common import JINJA_ENV
+from utils import pipeline_logging
+from constants import bq_utils as bq_consts
+from resources import fields_for, CDM_TABLES
+import constants.cdr_cleaner.clean_cdr as cdr_consts
+from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
+
+LOGGER = logging.getLogger(__name__)
+
+SANDBOX_QUERY = JINJA_ENV.from_string("""
+CREATE OR REPLACE TABLE `{{project_id}}.{{sandbox_id}}.{{intermediary_table}}` AS (
+SELECT * FROM `{{project_id}}.{{dataset_id}}.{{cdm_table}}`
+WHERE 
+    (coalesce({{date_fields}}) < DATE("{{cutoff_date}}")) 
```

Comment:
So this would check the first non-null date and non-null datetime field for each row. Is the first non-null field reliably always the correct one to use?

Reply:
All the date and DATETIME fields of each OMOP table (excluding the `person` table) are retrieved from the  snippet below and then rendered as the `{{date_fields}}` or `{{datetime_fields}}` in the query:

``` 
  # gets all fields from the affected table
  fields = fields_for(table)

  date_fields = []
  datetime_fields = []

  for field in fields:
      # appends only the date columns to the date_fields list
      if field['type'] in ['date']:
          date_fields.append(field['name'])

      # appends only the datetime columns to the datetime_fields list
      if field['type'] in ['timestamp']:
          datetime_fields.append(field['name'])
```

The `fields_for` function returns all the fields for the given table. For each field in the table if it has a `date` or `timestamp` `type`, it will append it to a `date_fields` or `datetime_fields` list. The contents of those lists are what is passed in as the `{{date_fields}}` and `{{datetime_fields}}` of the `SANDBOX_QUERY`

Let me know if that made sense, I'm happy to hop on a call if you need more explanation ☺️☺️

Reply:
@dcampbell-vumc looking at the query again, It is MAX(Date fields/datetime fields) we should be using instead of coalesce.

Reply:
@ksdkalluri can you explain this change? As it stands, it doesn't work in big query. I'm confused because instead of having separate lists of `date_fields` and `datetime_fields` this has it in one and I don't think we can compare `datetime` fields with a `date` type cutoff date.

Reply:
Like we discussed it's GREATEST() we have to use instead of MAX().